File: code/496.next-greater-element-i.java
/*
 * @lc app=leetcode id=496 lang=java
 *
 * [496] Next Greater Element I
 */

// @lc code=start

import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums1.length; i++) {
            map.put(nums1[i], i);
        }
        int[] res = new int[nums1.length];
        Arrays.fill(res, -1);
        Deque<Integer> stack = new LinkedList<>();
        for (int i = 0; i < nums2.length; i++) {
            while (!stack.isEmpty() && nums2[i] > nums2[stack.peek()]) {
                int top = nums2[stack.pop()];
                if (map.containsKey(top)) {
                    res[map.get(top)] = nums2[i];
                }
            }
            stack.push(i);
        }
        return res;
    }
}
// @lc code=end

File: code/496.next-greater-element-i.go
/*
 * @lc app=leetcode id=496 lang=golang
 *
 * [496] Next Greater Element I
 */

// @lc code=start
func nextGreaterElement(nums1 []int, nums2 []int) []int {
	res := make([]int, len(nums1))
	m := map[int]int{}
	for i, v := range nums1 {
		m[v] = i
		res[i] = -1
	}

	stack := []int{}

	for i, v := range nums2 {
		for len(stack) != 0 && v > nums2[stack[len(stack)-1]] {
			top := nums2[stack[len(stack)-1]]
			stack = stack[:len(stack)-1]
			if _, ok := m[top]; ok {
				res[m[top]] = v
			}
		}
		stack = append(stack, i)
	}
	return res
}

// @lc code=end

